src/adc.rs

```diff
@@ -0,0 +1,535 @@
+                pub fn setup_oneshot(&mut self) {
+                    // stop and clear overrun events
+                    self.rb.cr.modify(|_, w| w.adstp().set_bit());
+                    self.rb.isr.modify(|_, w| w.ovr().clear_bit());
+
+                    self.rb.cfgr.modify(|_, w| w
+                        .cont().clear_bit()
+                        .ovrmod().clear_bit()
+                    );
+
+                    self.rb.sqr1.modify(|_, w| w.l().bits(0));
+
+                    self.operation_mode = Some(OperationMode::OneShot);
+                }
+
+                fn set_align(&self, align: Align) {
+                    self.rb.cfgr.modify(|_, w| w.align().bit(align.bitvalue()));
+                }
+
+                fn enable(&mut self) {
+                    self.rb.cr.modify(|_, w| w.aden().set_bit());
+                    while self.rb.isr.read().adrdy().bit_is_clear() {}
+                }
+
+                fn disable(&mut self) {
+                    self.rb.cr.modify(|_, w| w.aden().clear_bit());
+                }
+
+
+                /// Calibrate according to 15.3.8 in the Reference Manual
+                fn calibrate(&mut self) {
+                    if !self.rb.cr.read().advregen().is_enabled() {
+                        self.advregen_enable();
+                        self.wait_advregen_startup();
+                    }
+
+                    self.disable();
+
+                    self.rb.cr.modify(|_, w| w
+                        // NOTE: needs to be adopted if implementing differential input
+                        .adcaldif().clear_bit()
+                        .adcal()   .set_bit());
+
+                    while self.rb.cr.read().adcal().bit_is_set() {}
+                }
+
+
+                fn wait_adc_clk_cycles(&self, cycles: u32) {
+                    let adc_clk_cycle = self.clocks.hclk().0 / (self.prescale as u32);
+                    asm::delay(adc_clk_cycle * cycles);
+                }
+
+                fn advregen_enable(&mut self){
+                    // need to go through intermediate first
+                    self.rb.cr.modify(|_, w| w.advregen().intermediate());
+                    self.rb.cr.modify(|_, w| w.advregen().enabled());
+
+                }
+
+                /// wait for the advregen to startup.
+                ///
+                /// This is based on the MAX_ADVREGEN_STARTUP_US of the device.
+                fn wait_advregen_startup(&self) {
+                        asm::delay((MAX_ADVREGEN_STARTUP_US * 1_000_000) / self.clocks.sysclk().0);
+                }
+
+                /// busy ADC read
+                fn convert_one(&mut self, chan: u8) -> u16 {
+                    self.ensure_oneshot();
+                    self.set_chan_smps(chan, SampleTime::default());
+                    self.select_single_chan(chan);
+
+                    self.rb.cr.modify(|_, w| w.adstart().set_bit());
+                    while self.rb.isr.read().eos().bit_is_clear() {}
```

Comment:
The reference (15.3.24) mentions that the "EOS flag is cleared by the software by writing 1 to it". There doesn't seem to be another mechanism to clear it, so I think we need to clear it here to ensure the next read works as expected.

Reply:
Good catch. I didn't notice it when testing on hardware as my brain is not fast enough to actually see a difference ;)

Reply:
Is this comment addressed?

Reply:
hmm it got lost in the rebase action - anyhow I like the new solution (https://github.com/stm32-rs/stm32f3xx-hal/pull/47/commits/efe8f23ced863e97c18b670a545722d6b82f5557) more